Montreal Forced Aligner tutorial (#7007)
galaxyproject/training-material: topics/digital-humanities/tutorials/mfa_tutorial/tutorial.md

Reviewer: i think I added a line too much here. One "questions" should be deleted, sorry

---
layout: tutorial_hands_on

title: Montreal Forced Aligner

> zenodo_link: https://zenodo.org/records/21371207

Reviewer: ```suggestion
zenodo_link: https://zenodo.org/records/21371207
questions:
  - How can you use Montreal Forced Aligner to align an audio and a TextGrid file in Galaxy?
```
This question part is mandatory, so it is currently shown as empty, so I would suggest adding a broad question here.

objectives:
- Use the Montreal Forced Aligner to break down speech into its smallest sounds and align them with their corresponding transcription.
time_estimation: 1H

> contributors:

Reviewer: ```suggestion
contributions:
  authorship:
    - wallajess
   reviewing:
    - Sch-Da
  testing:
    - Sch-Da
```

This section was expanded since the shell tutorial was created:
we can now differentiate different tasks for more detailed contributions.
If you put your github handle, this will automatically go to your hall of fame.
And this is also, why one of the tests was failing :)

> - Jessica Wallace

Reviewer: ```suggestion
```

Reviewer: The bot that checks this expects your github handle here. I added my suggestion below.

---

# Introduction

<!-- This is a comment. -->

This tutorial explains how to use the Montreal Forced Aligner (MFA) {% cite mcauliffe17_interspeech %} on the Galaxy platform. 

The aligner breaks down speech into its smallest possible sounds (phones) and aligns the indivdual 
sounds to their corresponding orthographic transcription. For example, the word 'fox' is broken down

> into four sounds: F AA K S or /f ɑː k s/ (depending on whether you use ARPAbet or IPA). The phonetic transcription of 

Reviewer: ```suggestion
into four sounds: F AA K S or /f ɑː k s/ (depending on whether you use ARPAbet or IPA for the transcription). The phonetic transcription of 
```

prefix to the filename (e.g. JW_interview.wav, JW_interview.TextGrid), then you 
can instruct the aligner to use the corresponding number of characters (2 in the previous example) as the speaker information. This speaker ID will be retained in the output, allowing you to later identify and analyze speech patterns by speaker.

These should then be compressed into one .zip file to be uploaded into Galaxy.

## Data Validation Checklist

Before uploading, verify your files meet these requirements:

- {% icon param-check %} Audio file and TextGrid have identical base filenames (e.g., `JW_interview.wav` and `JW_interview.TextGrid`)
- {% icon param-check %} Audio format is WAV or MP3 (WAV is best for quality)
- {% icon param-check %} TextGrid contains a single tier with short intervals, each representing one breath group
- {% icon param-check %} Transcription text matches the actual spoken content (typos will cause alignment errors)
- {% icon param-check %} Speaker ID prefix is consistent (e.g. all files start with 2-3 character speaker code)
- {% icon param-check %} Files are compressed into a single .zip archive

> 

Reviewer: ```suggestion
The following example file matches the above criteria and will be used in the next steps.
```
I thought it might be helpful to make sure those requirements are met in your example data set - so it can also be used as a reference hopefully?

## Uploading the data
> <hands-on-title> Data Upload </hands-on-title>
>
> 1. Create a new history for this tutorial
> 2. Import your files or if you want to follow the tutorial use the files from [Zenodo]({{ page.zenodo_link }}) or from
>    the shared data library (`GTN - Material` -> `{{ page.topic_name }}`
>     -> `{{ page.title }}`):
>
>    ```

> >    https://zenodo.org/api/records/20640646/files/the_fox_and_the_grapes.zip

Reviewer: ```suggestion
>    https://zenodo.org/records/21371207/files/the_fox_and_the_grapes.zip?download
```
The given link sends me to the page overview but not to the 2 datasets. I checked with the adjusted link and that worked. But maybe double check I am not in the wrong item altogether...

Author: That looks right now!

>    
>    ```
>    
>
>    {% snippet faqs/galaxy/datasets_import_via_link.md %}
>
>    {% snippet faqs/galaxy/datasets_import_from_data_library.md %}
>

> > 3. Rename the datasets to a more meaningful name(optional):

Reviewer: ```suggestion
> 3. Rename the datasets to a more meaningful name (optional):
```

> >    - Click on the dataset produced by MFA Find OOVs

Reviewer: ```suggestion
>    - Click on the dataset you uploaded
```
OOV come in as the next step only, right?

>    - Click the {% icon galaxy-pencil %} (edit) icon

> >    - Choose a meaningful name

Reviewer: ```suggestion
>    - Choose a meaningful name, for example, `the_fox_and_the_grapes.zip`
```

>    
> 4. Change the datatype of the dataset to mfa_corpus_model.zip
>
>    {% snippet faqs/galaxy/datasets_change_datatype.md datatype="datatypes" %}
>
{: .hands_on}

# Align the files

Before we align the files, we need to first check whether all of the words are 
in our dictionary. MFA offers the Find OOVs (out of vocabulary words) tool to do this.

## Sub-step with **MFA Find OOVs**

> > <hands-on-title> Task description </hands-on-title>

Reviewer: ```suggestion
> <hands-on-title> Find out-of-vocabulary-words </hands-on-title>
```
The general description should be adapted for a clearer headline

>

> > 1. {% tool [MFA Find OOVs](mfa_find_oovs) %} with the following parameters:

Reviewer: @wallajess I am currently investigating those messages

> >    - {% icon param-file %} *"Corpus Archive"*: `output` (Input dataset)

Reviewer: ```suggestion
>    - {% icon param-file %} *"Corpus Archive"*: `the_fox_and_the_grapes.zip` (Uploaded dataset)
```

>    - *"Dictionary Source"*: `Use a built-in MFA Dictionary`

> >        - *"Select Dictionary"*: `English Us (ARPA)`

Reviewer: ```suggestion
>        - *"Select Dictionary"*: `English Us (ARPA)`
>- *"Speaker Characters"*: `0`
```
If defaults are not changed, they do not show up in the training skeleton. For users, that can be un-intuitive so I would suggest adding the default parameters here

>
> 2. Rename the output file for clarity:
>    - Click on the dataset produced by MFA Find OOVs
>    - Click the {% icon galaxy-pencil %} (edit) icon
>    - Change the name to `OOVs_english_us` or similar
>    - This will help you identify this as the out-of-vocabulary words file
>

> > **Expected output:** A .zip file containing a text file (`oovs_found_english_us_arpa.txt`) listing all words from your transcription that are not in the English US ARPA dictionary. If no OOVs are found, the file will be empty (if that is the case, you can skip the next three steps and go straight to MFA Align).

Reviewer: really nicely added - thank you! I found this much clearer than in the first version.

>
{: .hands_on}

## Sub-step with **Unzip**

In order to generate the  pronunciations of the OOVs, we need the single file oovs_found_english_us_arpa.txt. We therefore need to unzip the output of Find OOVs so we can work with the file. The missing pronunciations can be generated automatically in the next step (MFA G2P) or manually using the editor.

> > <hands-on-title> Task description </hands-on-title>

Reviewer: ```suggestion
> <hands-on-title> Unzipping your file </hands-on-title>
```
The task description should be changed into something more meaningful. Mine is just a suggestion.

>

> > 1. {% tool [Unzip](toolshed.g2.bx.psu.edu/repos/imgteam/unzip/unzip/6.0+galaxy5) %} with the following parameters:

Reviewer: ```suggestion
> 1. {% tool [Unzip a file](toolshed.g2.bx.psu.edu/repos/imgteam/unzip/unzip/6.0+galaxy5) %} with the following parameters:
```
There are 2 unzip tools, I hope this clarifies for users, if they are searching directly in the tool bar

Author: Thanks, yes that makes a lot of sense!

>    - {% icon param-file %} *"Input file"*: `output_archive` (output of **MFA Find OOVs** {% icon tool %})
>    - *"What to extract"*: `Single file`
>        - *"File path"*: `oovs_found_english_us_arpa.txt`
>

> 2. Rename the output file for clarity:

Reviewer: ```
> 2. Rename the output file for clarity:
```
I think there is a formatting error here that breaks the "hands on" box in the rendering.

>    - Click on the dataset produced by MFA Find OOVs
>    - Click the {% icon galaxy-pencil %} (edit) icon
>    - Change the name to `unknown_words` or similar
>    - This will help you identify this as the out-of-vocabulary words file
>
>
>    > <comment-title> Check the OOVs </comment-title>
>    >
>    > Double-check the OOVs found file and make sure that none of the OOVs are typos. To do so, click on the {% icon galaxy-eye %} icon next to the dataset name to look at the file. If you do find a typo, you must correct this in the transcription, then upload the .TextGrid again and rerun the previous steps.

> >    > {: .comment}

Reviewer: I think, since this will be an issue below, there is also one > too much when closing the comment here. But please check if it properly renders without that.
```suggestion
>   {: .comment}
```

> 

Reviewer: The empty line here should also be deleted. This also breaks the Hands-on box- rendering

>
{: .hands_on}

## Sub-step with **MFA G2P**

Now we can generate pronunciations for each of the out of vocabulary words using MFA's built in grapheme-to-phoneme (G2P) tool. Select the same model you used for finding OOVs.

> > <hands-on-title> Task description </hands-on-title>

Reviewer: ```suggestion
> <hands-on-title> Generate pronunciations with G2P </hands-on-title>
```

>
> 1. {% tool [MFA G2P](mfa_g2p) %} with the following parameters:
>    - *"Input Mode"*: `Word List (Text File)`

> >        - {% icon param-file %} *"Word List"*: `unzipped_single` (output of **Unzip** {% icon tool %})

Reviewer: ```suggestion
>        - {% icon param-file %} *"Word List"*: `unkown_words` (output of **Unzip** {% icon tool %})
```

>    - *"G2P Model Source"*: `Use a built-in MFA Model`
>        - *"Select G2P Model"*: `English Us (ARPA)`
>
>    > <comment-title> Check the pronunciations manually </comment-title>
>    >
>    > It is important to check the pronunciations manually. MFA can also offer multiple possible pronunciations for each word. Correct any incorrect pronunciations and delete any additional pronunciations that do not apply. To do so, you can use the editor which can be found in Visualization on the side bar. There you can select the generated dictionary and correct the pronunciation.

> >    > {: .comment}

Reviewer: ```suggestion
>   {: .comment}
```
This box does not render properly - I assume because of one extra >
This is the case for some boxes below. But the one on praat works, so it is a syntax issue.

>
>    > <tip-title> Phonetic notation </tip-title>
>    >
>    > This tutorial uses ARPAbet phonetic notation for US English (e.g., AE, IH, K). Other models use IPA, which is the international standard for most other languages and varieties.

> >    > {: .tip}

Reviewer: ```suggestion
>    {: .tip}
```
I think it is the same here as the rendering of the box fails

>
> 2. Rename the output file for clarity:
>    - Click on the `output_dictionary` dataset produced by MFA G2P
>    - Click the {% icon galaxy-pencil %} (edit) icon
>    - Change the name to `generated_pronunciations_oovs` or `generated_dictionary`
>    - This clearly identifies this as the file containing generated pronunciations for OOV words
>
> **Expected output:** A dictionary file in MFA format (one entry per line: word followed by phonetic transcription). File size depends on the number of OOVs.
>
{: .hands_on}

## Sub-step with **MFA Merge**

Now we will add the new words to the built-in dictionary so that they are recognized when we run the aligner.

> > <hands-on-title> Task description </hands-on-title>

Reviewer: ```suggestion
> <hands-on-title> Add new words to your dictionary </hands-on-title>
```

>
> 1. {% tool [MFA Merge](mfa_merge) %} with the following parameters:
>    - *"Dictionary Source"*: `Use a built-in MFA Dictionary`
>        - *"Select Dictionary"*: `English Us (ARPA)`
>    - In *"Dictionary to add"*:
>        - {% icon param-repeat %} *"Insert Dictionary to add"*

> >            - {% icon param-file %} *"MFA dictionary"*: `output_dictionary` (output of **MFA G2P** {% icon tool %})

Reviewer: ```suggestion
>            - {% icon param-file %} *"MFA dictionary"*: `generated_pronunciations_oovs` (output of **MFA G2P** {% icon tool %})
```
Since you renamed this before, it might help taking that name up again, if I chose the correct one?

>
>
>    > <comment-title> If corrections were uploaded </comment-title>
>    >
>    > If you uploaded a file with corrected pronunciations, you will have to select the uploaded file with the corrected dictionary entries as your dictionary to add.

> >    > {: .comment}

Reviewer: ```suggestion
>   {: .comment}
```
This box is not rendered properly. I think it is due to one extra > but please check, if this is correct, if you have the capacity.

>
> 2. Rename the output file for clarity:
>    - Click on the `output` dataset produced by MFA Merge
>    - Click the {% icon galaxy-pencil %} (edit) icon
>    - Change the name to `updated_dictionary_english_us` or `merged_dictionary`
>    - This clearly identifies this as the merged dictionary containing both built-in entries and newly generated pronunciations
>
> **Expected output:** A merged dictionary file (single file, not a .zip) containing all entries from the built-in English US ARPA dictionary plus your newly generated pronunciations for OOV words. This is the complete dictionary your audio will be aligned against.
>
{: .hands_on}

## Sub-step with **MFA Align**

Now we're ready to align the pronunciations with the audio. Since we only have one spekaer and no spekaer ID, we leave the number of spekaer characters at 0. If you have a spekaer ID at the start of your file name, then select the number of characters corresponding to the speaker ID.qq

> > <hands-on-title> Task description </hands-on-title>

Reviewer: ```suggestion
> <hands-on-title> Align Audio and Pronounciation </hands-on-title>
```

>
> 1. {% tool [MFA Align](mfa_align) %} with the following parameters:

> >    - {% icon param-file %} *"Corpus Archive"*: `output` (Input dataset)

Reviewer: The corpus archive is the initial zip file, right? Then maybe rename this as well?
```suggestion
>    - {% icon param-file %} *"Corpus Archive"*: `the_fox_and_the_grapes.zip` (Input dataset)
```

>    - *"Dictionary Source"*: `Upload from History`

> >        - {% icon param-file %} *"Dictionary File"*: `output` (Input dataset)

Reviewer: And this would be the updated dictionary?

```suggestion
>        - {% icon param-file %} *"Dictionary File"*: `updated_dictionary_english_us` (output of **MFA Merge** {% icon tool %})
```

>    - In Praat, open the original audio file (File → Open → Read from file...)
>    - Then open the corresponding TextGrid (Objects window → Read → Read from file...)
>    - Select both and click "View & Edit" to visualize the alignment
>
>
{: .hands_on}

# Next Steps and Use Cases

Once you have aligned TextGrids, you can:

- **Measure vowel formants:** Extract and analyze vowel characteristics from aligned vowel intervals. To do so, you can use new_fave.
- **Compare pronunciations:** Analyze variation across speakers or time
- **Assess speech rate:** Measure duration of phonemes or words to study articulation speed
- **Study consonant voicing:** Use the precise time boundaries to measure voice onset time (VOT)

# Conclusion

In this tutorial, we uploaded a sample audio file and transcription. 
We checked for any out-of-vocabulary words (OOVs) missing from our dictionary and generated the pronunciations using G2P. 
After checking and, if necessary, correcting these pronunciations, we merged the generated pronunciations with the built-in dictionary and ran the aligner on our files, producing aligned TextGrid files with precise phoneme-level time boundaries.

> You now have a fully aligned and labeled corpus ready for phonetic anal

Reviewer: ```suggestion
You now have a fully aligned and labeled corpus ready for phonetic analysis.
```

> zenodo_link: https://zenodo.org/records/21371207

Reviewer: ```suggestion
---
layout: tutorial_hands_on
title: Montreal Forced Aligner
level: Introductory
zenodo_link: https://zenodo.org/records/21371207
questions:
```

Reviewer: Could you try this one?
I know the bot keeps complaining but like this, I think we were missing the initial header of the tutorial
